src/aws-secrets-controller/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+#![warn(clippy::suspicious_else_formatting)]
+#![warn(clippy::suspicious_unary_op_formatting)]
+#![warn(clippy::mut_mutex_lock)]
+#![warn(clippy::print_literal)]
+#![warn(clippy::same_item_push)]
+#![warn(clippy::useless_format)]
+#![warn(clippy::write_literal)]
+#![warn(clippy::redundant_closure)]
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use aws_config::meta::region::RegionProviderChain;
+use aws_config::SdkConfig;
+use aws_sdk_secretsmanager::config::{Builder as SecretsManagerConfigBuilder, Region};
+use aws_sdk_secretsmanager::error::SdkError;
+use aws_sdk_secretsmanager::primitives::Blob;
+use aws_sdk_secretsmanager::types::{Filter, FilterNameStringType, Tag};
+use aws_sdk_secretsmanager::Client;
+use futures::stream::StreamExt;
+use mz_repr::GlobalId;
+use mz_secrets::{SecretsController, SecretsReader};
+use mz_sql::catalog::EnvironmentId;
+
+#[derive(Clone, Debug)]
+pub struct AwsSecretsController {
+    pub client: Client,
+    pub secret_name_prefix: String,
+    pub environment_id: EnvironmentId,
+    pub default_tags: BTreeMap<String, String>,
+}
+
+pub async fn load_sdk_config(region: String) -> SdkConfig {
+    let region_provider = RegionProviderChain::first_try(Region::new(region));
+    let mut config_loader = aws_config::from_env().region(region_provider);
+    if let Ok(endpoint) = std::env::var("AWS_ENDPOINT_URL") {
+        config_loader = config_loader.endpoint_url(endpoint);
+    }
+    config_loader.load().await
+}
+
+async fn load_secrets_manager_client(region: String) -> Client {
+    let sdk_config = load_sdk_config(region).await;
+    let sm_config = SecretsManagerConfigBuilder::from(&sdk_config).build();
+    Client::from_conf(sm_config)
+}
+
+impl AwsSecretsController {
+    pub async fn new(
+        environment_id: EnvironmentId,
+        default_tags: BTreeMap<String, String>,
+    ) -> Self {
+        AwsSecretsController {
+            client: load_secrets_manager_client(environment_id.cloud_provider_region().to_owned())
+                .await,
+            secret_name_prefix: format!("/user-managed/{}/", environment_id),
+            environment_id,
+            default_tags,
+        }
+    }
+    fn tags(&self) -> Vec<Tag> {
+        self.default_tags
+            .iter()
+            .map(|(key, value)| Tag::builder().key(key).value(value).build())
+            .collect()
+    }
+
+    fn kms_key_alias(&self) -> String {
+        // Do not change this, as it must match the code in the region-controller.
```

Comment:
Should we add a TODO here to factor this into a pub method that the region controller can import and use?

Reply:
Sure. I'll add a TODO.